### PR TITLE
Switch spheres to new hyper-dimensional ones

### PR DESCRIPTION
**Pull request by aprokop**

### CODE REVIEW

File: examples/access_traits/example_cuda_access_traits.cpp [hunk 1]
```diff
@@ -55,7 +55,7 @@ struct ArborX::AccessTraits<Spheres, ArborX::PredicatesTag>
   static KOKKOS_FUNCTION auto get(Spheres const &d, std::size_t i)
   {
     return ArborX::intersects(
-        ArborX::Sphere{{{d.d_x[i], d.d_y[i], d.d_z[i]}}, d.d_r[i]});
+        ArborX::Sphere{ArborX::Point{d.d_x[i], d.d_y[i], d.d_z[i]}, d.d_r[i]});
```
Comment (masterleinad) [inline examples/access_traits/example_cuda_access_traits.cpp:58]: Much nicer!

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 2]
```diff
@@ -110,8 +110,7 @@ struct approx_expand_by_radius<PointTag, Point>
   {
     constexpr int DIM = GeometryTraits::dimension_v<Box>;
     using Coordinate = GeometryTraits::coordinate_type_t<Point>;
-    return ExperimentalHyperGeometry::Sphere<DIM, Coordinate>{
-        Kokkos::bit_cast<::ArborX::Point<DIM, Coordinate>>(point), r};
+    return Sphere{Kokkos::bit_cast<::ArborX::Point<DIM, Coordinate>>(point), r};
```
Comment (masterleinad) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:113]: Is the `bit_cast` still necessary here?

File: src/details/ArborX_NeighborList.hpp [hunk 3]
```diff
@@ -39,8 +39,7 @@ struct NeighborListPredicateGetter
 
     auto const &hyper_point =
         reinterpret_cast<::ArborX::Point<dim, Coordinate> const &>(pair.value);
```
Comment (masterleinad) [inline src/details/ArborX_NeighborList.hpp:41]: Do you still need the `reinterpret_cast`?

File: src/details/ArborX_PredicateHelpers.hpp [hunk 4]
```diff
@@ -173,8 +173,7 @@ struct AccessTraits<Experimental::PrimitivesWithRadius<Primitives>,
     // point structure (e.g., struct MyPoint { float y; float x; })
     auto const &hyper_point =
         reinterpret_cast<::ArborX::Point<dim, Coordinate> const &>(point);
```
Comment (masterleinad) [inline src/details/ArborX_PredicateHelpers.hpp:175]: Same here.

File: src/ArborX_DBSCAN.hpp [hunk 5]
```diff
@@ -67,7 +66,7 @@ struct WithinRadiusGetter
     auto const &hyper_point =
         reinterpret_cast<::ArborX::Point<dim> const &>(pair.value);
```
Comment (masterleinad) [inline src/ArborX_DBSCAN.hpp:67]: Same here.

File: src/ArborX_DBSCAN.hpp [hunk 6]
```diff
@@ -119,9 +118,7 @@ struct AccessTraits<Details::PrimitivesWithRadiusReorderedAndFiltered<
     // point structure (e.g., struct MyPoint { float y; float x; })
     auto const &hyper_point =
         reinterpret_cast<::ArborX::Point<dim> const &>(point);
```
Comment (masterleinad) [inline src/ArborX_DBSCAN.hpp:120]: Same here.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 2 above)
Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:113]: Yes. This geometries may come from users. We do an implicit undocumented cast here as we don't have a well defined concept of a point. It's ok for now but will need to change in the future.

File: src/details/ArborX_NeighborList.hpp (hunk 3 above)
Comment (aprokop) [reply src/details/ArborX_NeighborList.hpp:41]: Could be replaced by `bit_cast` but the issue is same as above: user data.

File: src/details/ArborX_PredicateHelpers.hpp (hunk 4 above)
Comment (aprokop) [reply src/details/ArborX_PredicateHelpers.hpp:175]: User data.

**Comment by aprokop** [comment]
So, all the comments are about `reinterpret_cast`/`bit_cast`. The issues is that our algorithms require running algorithms or constructing ArborX geometry types from some other types. Often, these other types are provided by a user, and we convert them to something that internally we can digest. We could follow boost and explicitly construct internal data by using getters.

I think this is a larger discussion and is outside of the scope of this PR.

**Pull request merged**